Unable to store rating curve with a single rating

Error storing a single rating curve to the database:

```
14-Jul-2023 03:08:54.869 SEVERE [https-jsse-nio-8443-exec-589] cwms.cda.api.RatingController.create -5127655873716502772: Failed to process create request
	hec.data.RatingException: Text is not a valid compressed or uncompressed CWMS Ratings XML instance.
		at mil.army.usace.hec.cwms.rating.io.xml.RatingXmlFactory.ratingSet(RatingXmlFactory.java:294)
		at cwms.cda.api.RatingController.deserializeFromXml(RatingController.java:164)
		at cwms.cda.api.RatingController.deserializeRatingSet(RatingController.java:152)
		at cwms.cda.api.RatingController.deserializeRatingSet(RatingController.java:144)
		at cwms.cda.api.RatingController.deserializeRatingSet(RatingController.java:138)
		at cwms.cda.api.RatingController.create(RatingController.java:111)
		at io.javalin.apibuilder.CrudFunction$3.invoke$lambda-0(CrudHandler.kt:32)
		at io.javalin.apibuilder.CrudFunctionHandler.handle(CrudHandler.kt)
		at cwms.cda.security.KeyAccessManager.manage(KeyAccessManager.java:36)
		at cwms.cda.security.MultipleAccessManager.manage(MultipleAccessManager.java:41)
		at io.javalin.http.JavalinServlet.addHandler$lambda-0(JavalinServlet.kt:96)
		at io.javalin.http.JavalinServlet$lifecycle$2$1$1.invoke(JavalinServlet.kt:43)
		at io.javalin.http.JavalinServlet$lifecycle$2$1$1.invoke(JavalinServlet.kt:43)
		at io.javalin.http.JavalinServletHandler.executeNextTask(JavalinServletHandler.kt:99)
		at io.javalin.http.JavalinServletHandler.queueNextTaskOrFinish$lambda-1(JavalinServletHandler.kt:85)
		at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:995)
		at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2137)
		at io.javalin.http.JavalinServletHandler.queueNextTaskOrFinish$javalin(JavalinServletHandler.kt:85)
		at io.javalin.http.JavalinServletHandler.executeNextTask$lambda-11$lambda-10(JavalinServletHandler.kt:119)
		at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:616)
		at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:628)
		at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:1996)
		at io.javalin.http.JavalinServletHandler.executeNextTask(JavalinServletHandler.kt:119)
		at io.javalin.http.JavalinServletHandler.queueNextTaskOrFinish$lambda-1(JavalinServletHandler.kt:85)
		at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:995)
		at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2137)
		at io.javalin.http.JavalinServletHandler.queueNextTaskOrFinish$javalin(JavalinServletHandler.kt:85)
		at io.javalin.http.JavalinServlet.service(JavalinServlet.kt:89)
		at javax.servlet.http.HttpServlet.service(HttpServlet.java:779)
		at cwms.cda.ApiServlet.service(ApiServlet.java:510)
		at javax.servlet.http.HttpServlet.service(HttpServlet.java:779)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)
		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
		at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
		at org.apache.catalina.filters.CorsFilter.handleNonCORS(CorsFilter.java:357)
		at org.apache.catalina.filters.CorsFilter.doFilter(CorsFilter.java:176)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
		at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197)
		at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
		at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:660)
		at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135)
		at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
		at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:687)
		at org.apache.catalina.authenticator.SingleSignOn.invoke(SingleSignOn.java:261)
		at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
		at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:360)
		at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:399)
		at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
		at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:893)
		at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1789)
		at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
		at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
		at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
		at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
		at java.lang.Thread.run(Thread.java:750)
		Suppressed: java.lang.IllegalArgumentException: Length of input is not multiple of 4
			at hec.util.Base64.decode(Base64.java:122)
			at hec.util.Base64.decode(Base64.java:168)
			at hec.util.Base64.decode(Base64.java:59)
			at hec.util.Base64.decode(Base64.java:49)
			at hec.util.TextUtil.uncompress(TextUtil.java:422)
			at mil.army.usace.hec.cwms.rating.io.xml.RatingXmlFactory.ratingSet(RatingXmlFactory.java:289)
			... 57 more
	Caused by: hec.data.RatingException: java.lang.NullPointerException
		at mil.army.usace.hec.cwms.rating.io.xml.RatingSetContainerXmlFactory.parseInputSource(RatingSetContainerXmlFactory.java:346)
		at mil.army.usace.hec.cwms.rating.io.xml.RatingSetContainerXmlFactory.parseReader(RatingSetContainerXmlFactory.java:316)
		at mil.army.usace.hec.cwms.rating.io.xml.RatingSetContainerXmlFactory.ratingSetContainerFromXml(RatingSetContainerXmlFactory.java:247)
		at mil.army.usace.hec.cwms.rating.io.xml.RatingSetContainerXmlFactory.ratingSetContainerFromXml(RatingSetContainerXmlFactory.java:237)
		at mil.army.usace.hec.cwms.rating.io.xml.RatingXmlFactory.ratingSet(RatingXmlFactory.java:285)
		... 57 more
	Caused by: java.lang.NullPointerException
		at mil.army.usace.hec.cwms.rating.io.xml.RatingSetContainerXmlFactory.endElement(RatingSetContainerXmlFactory.java:1465)
		at org.xml.sax.helpers.ParserAdapter.endElement(ParserAdapter.java:652)
		at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.endElement(AbstractSAXParser.java:602)
		at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanEndElement(XMLDocumentFragmentScannerImpl.java:1784)
		at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2969)
		at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
		at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:113)
		at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:507)
		at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:867)
		at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:796)
		at com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:142)
		at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1216)
		at org.xml.sax.helpers.ParserAdapter.parse(ParserAdapter.java:431)
		at org.xml.sax.helpers.XMLFilterImpl.parse(XMLFilterImpl.java:357)
		at mil.army.usace.hec.cwms.rating.io.xml.RatingSetContainerXmlFactory.parseInputSource(RatingSetContainerXmlFactory.java:338)
		... 61 more
```


The rating xml sent to CDA is the following:

```
<?xml version="1.0" encoding="utf-8"?>
<ratings
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="http://www.hec.usace.army.mil/xmlSchema/cwms/Ratings.xsd">
    <simple-rating office-id="SWT">
        <rating-spec-id>07179500.Stage;Flow.EXSA.PRODUCTION</rating-spec-id>
        <units-id>ft;cfs</units-id>
        <effective-date>2023-07-14T03:10:00Z</effective-date>
        <transition-start-date/>
        <create-date>2023-07-14T03:10:00Z</create-date>
        <active>true</active>
        <description>NEOSHO R AT COUNCIL GROVE, KS Expanded, Shift-Adjusted PRODUCTION Stream Rating</description>
        <rating-points>
            <point>
                <ind>0.0</ind>
                <dep>0.0</dep>
            </point>
            <point>
                <ind>1.0</ind>
                <dep>2.0</dep>
            </point>
        </rating-points>
    </simple-rating>
</ratings>
```

--- a/cwms-data-api/src/main/java/cwms/cda/api/RatingController.java
+++ b/cwms-data-api/src/main/java/cwms/cda/api/RatingController.java
@@ -42,7 +42,12 @@ import io.javalin.core.util.Header;
 import io.javalin.core.validation.JavalinValidation;
 import io.javalin.http.Context;
 import io.javalin.http.HttpCode;
-import io.javalin.plugin.openapi.annotations.*;
+import io.javalin.plugin.openapi.annotations.HttpMethod;
+import io.javalin.plugin.openapi.annotations.OpenApi;
+import io.javalin.plugin.openapi.annotations.OpenApiContent;
+import io.javalin.plugin.openapi.annotations.OpenApiParam;
+import io.javalin.plugin.openapi.annotations.OpenApiRequestBody;
+import io.javalin.plugin.openapi.annotations.OpenApiResponse;
 import mil.army.usace.hec.cwms.rating.io.xml.RatingXmlFactory;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
@@ -108,7 +113,7 @@ public class RatingController implements CrudHandler {
         try (final Timer.Context ignored = markAndTime("create"); DSLContext dsl =
                 getDslContext(ctx)) {
             RatingDao ratingDao = getRatingDao(dsl);
-            RatingSet ratingSet = deserializeRatingSet(ctx);
+            String ratingSet = deserializeRatingSet(ctx);
             boolean storeTemplate = ctx.queryParamAsClass(STORE_TEMPLATE, Boolean.class).getOrDefault(true);
             ratingDao.create(ratingSet, storeTemplate);
             ctx.status(HttpServletResponse.SC_ACCEPTED).json("Created RatingSet");
@@ -134,24 +139,18 @@ public class RatingController implements CrudHandler {
         return contentType;
     }
 
-    private RatingSet deserializeRatingSet(Context ctx) throws IOException, RatingException {
-        return deserializeRatingSet(ctx.body(), getContentType(ctx));
+    private String deserializeRatingSet(Context ctx) throws IOException, RatingException {
+        return deserializeRatingSet(ctx.body(), getContentType(ctx).getType());
     }
 
-
-    public RatingSet deserializeRatingSet(String body, ContentType contentType)
-            throws IOException, RatingException {
-        return deserializeRatingSet(body, contentType.getType());
-    }
-
-    public RatingSet deserializeRatingSet(String body, String contentType) throws IOException,
+    String deserializeRatingSet(String body, String contentType) throws IOException,
             RatingException {
-        RatingSet retval;
+        String retval;
 
         if ((Formats.XML).equals(contentType)) {
-            retval = deserializeFromXml(body);
+            retval = body;
         } else if ((Formats.JSON).equals(contentType)) {
-            retval = JsonRatingUtils.fromJson(body);
+            retval = RatingXmlFactory.toXml(JsonRatingUtils.fromJson(body), "");
         } else {
             throw new IOException("Unexpected format:" + contentType);
         }
@@ -446,7 +445,7 @@ public class RatingController implements CrudHandler {
             RatingDao ratingDao = getRatingDao(dsl);
 
             boolean storeTemplate = ctx.queryParamAsClass(STORE_TEMPLATE, Boolean.class).getOrDefault(true);
-            RatingSet ratingSet = deserializeRatingSet(ctx);
+            String ratingSet = deserializeRatingSet(ctx);
             ratingDao.store(ratingSet, storeTemplate);
             ctx.status(HttpServletResponse.SC_ACCEPTED).json("Updated RatingSet");
         } catch (IOException | RatingException ex) {

--- a/cwms-data-api/src/main/java/cwms/cda/data/dao/RatingDao.java
+++ b/cwms-data-api/src/main/java/cwms/cda/data/dao/RatingDao.java
@@ -36,7 +36,7 @@ public interface RatingDao {
 
     static final Pattern officeMatcher = Pattern.compile(".*office-id=\"(.*?)\"");
     
-    void create(RatingSet ratingSet, boolean storeTemplate) throws IOException, RatingException;
+    void create(String ratingSet, boolean storeTemplate) throws IOException, RatingException;
 
     RatingSet retrieve(RatingSet.DatabaseLoadMethod method, String officeId, String specificationId,
                        Instant start, Instant end) throws IOException, RatingException;
@@ -45,7 +45,7 @@ public interface RatingDao {
                            String start, String end, String timezone);
 
 
-    void store(RatingSet ratingSet, boolean storeTemplate) throws IOException, RatingException;
+    void store(String ratingSet, boolean storeTemplate) throws IOException, RatingException;
 
     void delete(String officeId, String specificationId, Instant start, Instant end);
 

--- a/cwms-data-api/src/main/java/cwms/cda/data/dao/RatingSetDao.java
+++ b/cwms-data-api/src/main/java/cwms/cda/data/dao/RatingSetDao.java
@@ -24,6 +24,9 @@
 
 package cwms.cda.data.dao;
 
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.dataformat.xml.XmlMapper;
 import hec.data.RatingException;
 import hec.data.cwmsRating.RatingSet;
 import mil.army.usace.hec.cwms.rating.io.jdbc.ConnectionProvider;
@@ -36,6 +39,7 @@ import java.io.IOException;
 import java.sql.Connection;
 import java.sql.Timestamp;
 import java.time.Instant;
+import java.util.List;
 
 public class RatingSetDao extends JooqDao<RatingSet> implements RatingDao {
 
@@ -45,13 +49,14 @@ public class RatingSetDao extends JooqDao<RatingSet> implements RatingDao {
     }
 
     @Override
-    public void create(RatingSet ratingSet, boolean storeTemplate) throws IOException, RatingException {
+    public void create(String ratingSetXml, boolean storeTemplate) throws IOException, RatingException {
         try {
             connection(dsl, c -> {
-                setOffice(c, ratingSet.getRatingSpec().getOfficeId());
                 // can't exist if we are creating, if it exists use store
-                boolean overwriteExisting = false;
-                RatingJdbcFactory.store(ratingSet, c, overwriteExisting, storeTemplate);
+                String office = extractOfficeId(ratingSetXml);
+                DSLContext context = getDslContext(c, office);
+                CWMS_RATING_PACKAGE.call_STORE_RATINGS_XML__5(context.configuration(),
+                        ratingSetXml, "T", storeTemplate ? "T" : "F");
             });
         } catch (DataAccessException ex) {
             Throwable cause = ex.getCause();
@@ -60,6 +65,18 @@ public class RatingSetDao extends JooqDao<RatingSet> implements RatingDao {
             }
             throw new IOException("Failed to create Rating", ex);
         }
+    }
+
+    private static String extractOfficeId(String ratingSet) throws JsonProcessingException {
+        XmlMapper xmlMapper = new XmlMapper();
+        JsonNode node = xmlMapper.readTree(ratingSet);
+        List<JsonNode> values = node.findValues("office-id");
+        String office = "";
+        if (!values.isEmpty()) {
+            //Getting the last instance since the order is template, spec, rating
+            office = values.get(values.size() - 1).textValue();
+        }
+        return office;
     }
 
     @Override
@@ -109,12 +126,13 @@ public class RatingSetDao extends JooqDao<RatingSet> implements RatingDao {
 
     // store/update
     @Override
-    public void store(RatingSet ratingSet, boolean includeTemplate) throws IOException, RatingException {
+    public void store(String ratingSetXml, boolean includeTemplate) throws IOException, RatingException {
         try {
             connection(dsl, c -> {
-                setOffice(c,ratingSet.getRatingSpec().getOfficeId());
-                boolean overwriteExisting = true;
-                RatingJdbcFactory.store(ratingSet, c, overwriteExisting, includeTemplate);
+                String office = extractOfficeId(ratingSetXml);
+                DSLContext context = getDslContext(c, office);
+                CWMS_RATING_PACKAGE.call_STORE_RATINGS_XML__5(context.configuration(),
+                        ratingSetXml, "F", includeTemplate ? "T" : "F");
             });
         } catch (DataAccessException ex) {
             Throwable cause = ex.getCause();

